contrail-provisioner/vrouternode/vrouternode.go

```diff
@@ -0,0 +1,223 @@
+	vrouterInfoLog.Printf("Creating %s %s", c.Hostname, nodeType)
+	gscObjects := []*contrailtypes.GlobalSystemConfig{}
+	gscObjectsList, err := contrailClient.List("global-system-config")
+	if err != nil {
+		return err
+	}
+
+	if len(gscObjectsList) == 0 {
+		vrouterInfoLog.Println("no gscObject")
+	}
+
+	for _, gscObject := range gscObjectsList {
+		obj, err := contrailClient.ReadListResult("global-system-config", &gscObject)
+		if err != nil {
+			return err
+		}
+		gscObjects = append(gscObjects, obj.(*contrailtypes.GlobalSystemConfig))
+	}
+	for _, gsc := range gscObjects {
+		virtualRouter := &contrailtypes.VirtualRouter{}
+		virtualRouter.SetVirtualRouterIpAddress(c.IPAddress)
+		virtualRouter.SetParent(gsc)
+		virtualRouter.SetName(c.Hostname)
+		annotations := contrailclient.ConvertMapToContrailKeyValuePairs(c.Annotations)
+		virtualRouter.SetAnnotations(&annotations)
+		if err := contrailClient.Create(virtualRouter); err != nil {
+			return err
+		}
+		return nil
+	}
+	return nil
+}
+
+// Update updates a VirtualRouter instance
+func (c *VrouterNode) Update(contrailClient contrailclient.ApiClient) error {
+	vrouterInfoLog.Printf("Updating %s %s", c.Hostname, nodeType)
+	obj, err := contrailclient.GetContrailObjectByName(contrailClient, string(nodeType), c.Hostname)
+	if err != nil {
+		return err
+	}
+	virtualRouter := obj.(*contrailtypes.VirtualRouter)
+	virtualRouter.SetVirtualRouterIpAddress(c.IPAddress)
+	annotations := contrailclient.ConvertMapToContrailKeyValuePairs(c.Annotations)
+	virtualRouter.SetAnnotations(&annotations)
+	return contrailClient.Update(virtualRouter)
+}
+
+// Delete deletes a VirtualRouter instance and it's vhost0 VirtualMachineInterfaces
+func (c *VrouterNode) Delete(contrailClient contrailclient.ApiClient) error {
+	vrouterInfoLog.Printf("Deleting %s %s", c.Hostname, nodeType)
+	obj, err := contrailclient.GetContrailObjectByName(contrailClient, string(nodeType), c.Hostname)
+	if err != nil {
+		return err
+	}
+	virtualRouter := obj.(*contrailtypes.VirtualRouter)
+	err = deleteVMIs(virtualRouter, contrailClient)
+	if err != nil {
+		return err
+	}
+	vrouterInfoLog.Println("Deleting VirtualRouter ", obj.GetName())
+	if err = contrailClient.Delete(obj); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (c *VrouterNode) GetHostname() string {
+	return c.Hostname
+}
+
+func (c *VrouterNode) GetAnnotations() map[string]string {
+	return c.Annotations
+}
+
+func (c *VrouterNode) SetAnnotations(annotations map[string]string) {
+	c.Annotations = annotations
+}
+
+func (c *VrouterNode) Equal(otherNode contrailnode.ContrailNode) bool {
+	otherVrouterNode, ok := otherNode.(*VrouterNode)
+	if !ok {
+		return false
+	}
+	return otherVrouterNode.Hostname == c.Hostname && otherVrouterNode.IPAddress == c.IPAddress &&
+		reflect.DeepEqual(otherVrouterNode.Annotations, c.Annotations)
+}
+
+func (c *VrouterNode) EnsureDependencies(contrailClient contrailclient.ApiClient) error {
```

Comment:
Do we really need this extra method in the interface that is only implemented by one class?
Maybe this could be called internally in create and update methods of the type that requires it?

Reply:
The problem is that it also should be called on the nodes that have `noop` action set by the createActionMap function. There may be a case that the "vhost0" virtual-machine-interface is deleted, but the virtual-router object is not modified, or for example that the virtual-router creation finishes successfully, but virtual-machine-interface times out. Currently, the EnsureDependencies will create the missing vmi on next pass.

Reply:
So `noop` action is actually do some some action

Reply:
Maybe it should method `EnsureDependencies` should be named `Noop`?

Reply:
Removed `EnsureDependencies` and `noop` action. Update is executed instead and vmi creation has been moved to vrouter's create and update methods.